### ISSUE TITLE
no-floating-promises didn't catch an "error" that I think it should have .. 

### ISSUE THREAD

**Issue by sjmcdowall** [palantir/tslint#4115]
### Bug Report

- __TSLint version__:  5.11.0
- __TypeScript version__:  2.9.2
- __Running TSLint via__: (pick one) CLI / Node.js API / VSCode / grunt-tslint / Atom / Visual Studio / etc
     via yarn script or within VSCode

#### TypeScript code being linted

```ts
async function passwordMatches(
  password: string,
  hashedPassword: string,
  hashType: EMemberHashType,
): Promise<boolean> {
  let doesPasswordMatch = false;
  ...const blah = await something();
  return doesPasswordMatch;
}

     // Do the basic password check
      if (!passwordMatches(password, member.password, hashType)) {
        authService.logger.trace(`user ${login} password did not match`);
        reject(new Error(INVALID_MESSAGE));
        return;
      }
```

with `tslint.json` configuration:

```json
{
  "defaultSeverity": "error",
  "extends": [
    "tslint:recommended",
    "tslint-consistent-codestyle"
  ],
  "rules": {
    "eofline": true,
    "indent": [true, "spaces", 2],
    "quotemark": [true, "avoid-escape", "single"],
    "ban-types": [true, ["Object", "Use {} instead."], ["String", "Use string instead"], ["Number", "Use number instead"]],
    "member-access": [true, "no-public"],
    "no-inferrable-types": [true, "ignore-params"],
    "import-blacklist": [true, "lodash"],
    "no-internal-module": true,
    "no-non-null-assertion": true,
    "no-parameter-reassignment": true,
    "no-reference": true,
    "no-var-requires": true,
    "promise-function-async": true,
    "unified-signatures": true,
    "await-promise": true,
    "ban-comma-operator": true,
    "curly": true,
    "no-arg": true,
    "no-bitwise": true,
    "no-conditional-assignment": true,
    "no-construct": true,
    "no-console": [true, "log", "error"],
    "no-debugger": true,
    "no-duplicate-super": true,
    "no-duplicate-switch-case": true,
    "no-duplicate-variable": true,
    "no-empty": true,
    "no-eval": true,
    "no-floating-promises": true,
    "no-for-in-array": true,
    "no-inferred-empty-object-type": true,
    "no-invalid-template-strings": true,
    "no-misused-new": true,
    "no-object-literal-type-assertion": true,
    "no-return-await": true,
    "no-sparse-arrays": true,
    "no-string-literal": true,
    "no-string-throw": true,
    "no-switch-case-fall-through": true,
    "no-this-assignment": [true, { "allow-destructuring": true }],
    "no-unnecessary-class": [true, "allow-constructor-only"],
    "no-unsafe-any": true,
    "no-unused": true,
    "no-unused-expression": true,
    "no-use-before-declare": true,
    "no-var-keyword": true,
    "prefer-object-spread": true,
    "radix": true,
    "restrict-plus-operands": true,
    "triple-equals": true,
    "use-isnan": true,
    "cyclomatic-complexity": [true, 15],
    "deprecation": true,
    "linebreak-style": [true, "LF"],
    "max-classes-per-file": [true, 1],
    "max-file-line-count": [true, 300],
    "max-line-length": [true, 160],
    "no-duplicate-imports": true,
    "prefer-const": true,
    "trailing-comma": [true, {"multiline": "always", "singleline": "never"}],
    "arrow-return-shorthand": true,
    "class-name": true,
    "import-spacing": true,
    "interface-name": [true, "always-prefix"],
    "new-parens": true,
    "no-consecutive-blank-lines": true,
    "no-irregular-whitespace": true,
    "no-trailing-whitespace": true,
    "no-unnecessary-callback-wrapper": true,
    "one-variable-per-declaration": true,
    "prefer-method-signature": true,
    "prefer-template": true,
    "semicolon": [true, "always"],
    "space-before-function-paren": [true, {"asyncArrow": "always", "named": "never"}],
    "variable-name": [true, "ban-keywords", "check-format", "allow-leading-underscore"],
    "whitespace": [
        true,
        "check-branch",
        "check-decl",
        "check-module",
        "check-separator",
        "check-operator",
        "check-typecast",
        "check-rest-spread",
        "check-type",
        "check-typecast",
        "check-type-operator",
        "check-preblock"
    ]
  }
}
```

#### Actual behavior

No errors or warnings 

#### Expected behavior

Some warning message about using a return of a Promise in a boolean expression?   This caused a large security hole not doing the await on the function ...

**Comment by JoshuaKGoldberg** [comment]
Hey @sjmcdowall - I'm a little confused about which line should have been an error. Is it the `!passwordMatches(...)`?

Also, have you tried `strict-boolean-expressions`?

**Comment by sjmcdowall** [comment]
Yes the if(!passwordMatches()) since passwordMatches() returns a Promise that is not being handled at all -- i.e. it's "floating" as far as I am concerned.  It obviously should have had an "await" but did not .. 

To honest I am not sure if I have strict-boolean-expressions or not since I also include:
   "tslint:recommended",
    "tslint-consistent-codestyle"

As seen above and don't know if they include that or not .. 

However, this is , to me, a "floating promise" error and to a certain point I think the fact it was used in an a boolean expression is a "side effect" and that the floating-promise should have caught it, not the fact it was a bad boolean?

**Comment by JoshuaKGoldberg** [comment]
https://github.com/palantir/tslint/blob/master/src/configs/recommended.ts

Ah, you don't have it enabled. It would have complained on that line, since the issue you described of expecting a boolean-like and receiving a Promise.

This is a dup of #3804 and #3983. +1 to this providing more evidence that we need better docs and a more configurable `strict-boolean-expressions` here 😅

**Comment by sjmcdowall** [comment]
OK -- I added the option, and yes, lo and behold I do get a linter error (oddly, not in VSCode just command line lint) with the message:
ERROR: /Users/sjm/work/mm/mms4/auth-server/src/lib/auth.ts[96, 12]: This type is not allowed in the operand for the '!' operator because it is always truthy. Only booleans are allowed.

Which although true, seems to me to miss the main (and pain point) here, which is that the Promise was not correctly "used" with an await or equivalent .. <shrug>

**Comment by JoshuaKGoldberg** [comment]
> not in VSCode just command line lint

Oh yes: https://github.com/Microsoft/vscode-tslint/issues/70

**Comment by johnwiseheart** [comment]
Closing since this appears to be a dup of #3804 and #3983.
